### PR TITLE
chore: Speed up docker builds a bit by reducing layer count.

### PR DESCRIPTION
**Pull request by iphydf**
This is especially noticeable on local builds (less so on the github workers).

- - -
This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2603)

### CODE REVIEW

**Comment by nurupo** [comment]
So, instead of copying the source files from the local git repo checkout into the CI Docker containers on every CI run, we have a single CI action that gathers all the source files from the local git repo checkout and uploads that as an image to Dockerhub, with all other CIs actions pulling that image back from the Dockerhub via the Internet? Are... you sure this speeds them up?

Not familiar with how the docker actions work: do PRs also publish `toxchat/c-toxcore:sources` to Dockerhub, or only git master pushes do? I would imagine the PRs to lack credentials to publish to Dockerhub, but not too sure how our CI is setup.

If it's git master pushes only that publish `toxchat/c-toxcore:sources` to Dockerhub, how do PRs then get the PR source code after pulling master branch's sources via `toxchat/c-toxcore:sources` from Dockerhub? I don't see the source code being updated to that of PRs anywhere.

Are there any race conditions? If we do push `toxchat/c-toxcore:sources` to Dockerhub in PRs, what if there are two PRs running CIs concurrently, so PR1 publishes `toxchat/c-toxcore:sources` to Dockerhub but then PR2 overwrites` toxchat/c-toxcore:sources` on Dockerhub and the rest of PR1 uses `toxchat/c-toxcore:sources` from PR2? Similar can happen on master branch pushes, though it's less of an issue if CI for push1 uses source code of push2 since push2 would include things from push1, however it would be bad if push2 runs using the source code from push1 instead.

Does Dockerhub invalidate the cache promptly? What if we publish an image to Dockerhub, but Dockerhub still serves the old image for the next X minutes because it's cached somewhere on a CDN?

**Review by nurupo — COMMENTED** [review]
So, instead of copying the source files from the local git repo checkout into the CI Docker containers on every CI run, we have a single CI action that gathers all the source files from the local git repo checkout and uploads that as an image to Dockerhub, with all other CIs actions pulling that image back from the Dockerhub via the Internet? Are... you sure this speeds them up?

Not familiar with how the docker actions work: do PRs also publish `toxchat/c-toxcore:sources` to Dockerhub, or only git master pushes do? I would imagine the PRs to lack credentials to publish to Dockerhub, but not too sure how our CI is setup.

If it's git master pushes only that publish `toxchat/c-toxcore:sources` to Dockerhub, how do PRs then get the PR source code after pulling master branch's sources via `toxchat/c-toxcore:sources` from Dockerhub? I don't see the source code being updated to that of PRs anywhere.

Are there any race conditions? If we do push `toxchat/c-toxcore:sources` to Dockerhub in PRs, what if there are two PRs running CIs concurrently, so PR1 publishes `toxchat/c-toxcore:sources` to Dockerhub but then PR2 overwrites` toxchat/c-toxcore:sources` on Dockerhub and the rest of PR1 uses `toxchat/c-toxcore:sources` from PR2? Similar can happen on master branch pushes, though it's less of an issue if CI for push1 uses source code of push2 since push2 would include things from push1, however it would be bad if push2 runs using the source code from push1 instead.

Does Dockerhub invalidate the cache promptly? What if we publish an image to Dockerhub, but Dockerhub still serves the old image for the next X minutes because it's cached somewhere on a CDN?

**Comment by nurupo** [comment]
Ok, I see where my confusion came from. I misread `push` as `publish`. It's also that each analysis tool does that, not just one job.

So, each analysis tool job gathers source files from the local git repo checkout and pushes and builds an image containing them, and then runs the tool-specific Docker build that uses that local image. No Dockerhub involved at all.

File: other/docker/misra/Makefile [hunk 1]
```diff
@@ -155,7 +155,12 @@ SUPPRESSIONS += 21.10
 
 CPPFLAGS := -DCMP_NO_FLOAT=1 -DMIN_LOGGER_LEVEL=LOGGER_LEVEL_TRACE
 
-SOURCES := $(shell find /src/workspace/c-toxcore -name "*.c")
+FIND_FLAGS := -name "*.c"								\
+			  -and -not -wholename "*/auto_tests/*"		\
+			  -and -not -wholename "*/other/*"			\
```
Comment (nurupo) [inline other/docker/misra/Makefile:160]: Doesn't that exclude the bootstrap daemon and other utilities when it didn't before? Do we not want to run misra on them?

File: other/docker/tokstyle/tokstyle.Dockerfile [hunk 2]
```diff
@@ -15,8 +16,11 @@ RUN apt-get update && \
...
 
-COPY toxav/ /src/workspace/c-toxcore/toxav/
-COPY toxcore/ /src/workspace/c-toxcore/toxcore/
-COPY toxencryptsave/ /src/workspace/c-toxcore/toxencryptsave/
-COPY third_party/cmp/cmp.h /src/workspace/c-toxcore/third_party/cmp/cmp.h
-RUN /bin/check-c $(find /src/workspace/c-toxcore -name "*.c")
+COPY --from=sources /src/ /src/workspace/c-toxcore/
+RUN /bin/check-c $(find /src/workspace/c-toxcore -name "*.c" \
+ -and -not -wholename "*/auto_tests/*" \
+ -and -not -wholename "*/other/*" \
```
Comment (nurupo) [inline other/docker/tokstyle/tokstyle.Dockerfile:22]: Doesn't that exclude the bootstrap daemon and other utilities when it didn't before? Do we not want to run check-c on them?

**Review by nurupo — CHANGES_REQUESTED** [review]

**Comment by nurupo** [comment]
Somewhat curious as about by how much does this speed up the local builds. It can't be more than 1-5 seconds on a modern computer with an ssd, can it? I would actually think that it makes them slower, since for every tool you run you now also run `other/docker/sources/build` which calls docker to build the source image, with the docker having to analyze the fs to see if any files have were modified/added/removed.

File: other/docker/misra/Makefile (hunk 1 above)
Comment (iphydf) [reply other/docker/misra/Makefile:160]: It does, and [no we don't](https://github.com/TokTok/c-toxcore/blob/d6d67d56f3e56d30cefd81f70e68c041a8c4bf2a/other/docker/misra/Dockerfile#L15-L19).

**Comment by nurupo** [comment]
LGFM, just want to get those conversations resolved (and possibly the speed-up question answered).

File: other/docker/tokstyle/tokstyle.Dockerfile (hunk 2 above)
Comment (iphydf) [reply other/docker/tokstyle/tokstyle.Dockerfile:22]: It does, and [no we don't](https://github.com/TokTok/c-toxcore/blob/d6d67d56f3e56d30cefd81f70e68c041a8c4bf2a/other/docker/tokstyle/Dockerfile#L18-L21).

**Comment by iphydf** [comment]
> Somewhat curious as by how much does this speed up the local builds. It can't be more than 1-5 seconds on a modern computer with an ssd, can it? I would actually think that it makes them slower, since for every tool you run you now also run `other/docker/sources/build` which calls docker to build the source image, with the docker having to analyze the fs to see if any files have were modified/added/removed.

Not everybody has SSD :). On SSD, it almost doesn't matter, but on HDD it's quite significant.

**Comment by iphydf** [comment]
It matters more when you run multiple of these docker builds in sequence, so you only need to build sources once.

**Review by nurupo — APPROVED** [review]

**Pull request merged**